tt_metal/hw/firmware/src/brisc.cc

```diff
@@ -354,43 +357,69 @@ int main() {
     // Wait for ncrisc to halt
     wait_for_ncrisc_to_halt();
 
-    mailboxes->launch.go.run = RUN_MSG_DONE;
+    mailboxes->go_message.signal = RUN_MSG_DONE;
 
     while (1) {
         init_sync_registers();
         reset_ncrisc_with_iram();
 
         WAYPOINT("GW");
-        while (mailboxes->launch.go.run != RUN_MSG_GO);
+
+        while (mailboxes->go_message.signal != RUN_MSG_GO) {
+            // While the go signal for kernel execution is not sent, check if the worker was signalled
+            // to reset its launch message read pointer.
+            if (mailboxes->go_message.signal == RUN_MSG_RESET_READ_PTR) {
+                // Set the rd_ptr on workers to specified value
+                mailboxes->launch_msg_rd_ptr = 0;
+                uint64_t dispatch_addr =
+                    NOC_XY_ADDR(NOC_X(mailboxes->go_message.master_x),
+                    NOC_Y(mailboxes->go_message.master_y), DISPATCH_MESSAGE_ADDR);
+                mailboxes->go_message.signal = RUN_MSG_DONE;
+                // Notify dispatcher that this has been done
+                DEBUG_SANITIZE_NOC_ADDR(noc_index, dispatch_addr, 4);
+                noc_fast_atomic_increment(
```

Comment:
for code bloat reduction, would be good to see if this can be conslidated as one call to noc w/ the code below.  I didn't think through the logic and if that would be a mess, but seems it should work ok?

Reply:
The reasoning here was to overlap overhead introduced by conditionals (checking if we have `RUN_MSG_GO` vs `RUN_MSG_RESET_READ_PTR`) with waiting for the go signal. We essentially wanted to avoid adding conditionals to the kernel execution section.

The way I see the 2 semaphore inc calls being consolidated is by doing something like this in the `while(1)` loop:

```
while ((go_message_signal = mailboxes->go_message.signal) != RUN_MSG_GO || (go_message_signal = mailboxes->go_message.signal) != RUN_MSG_READ_PTR) {
 // do nothing
}

if (go_message_signal == RUN_MSG_RESET_READ_PTR) {
 // reset launch message read ptr
} else {
 // execute kernel
}
```

This causes an extra check on the workers even if dispatch has already sent the go signal. I tried this approach and saw ~1-2% worse perf running empty kernels.